File: metasystem/internal/ui/web/_app/src/application/application.test.ts
import { describe, expect, it } from "vitest";

import type { Landed, Page, Problem } from "./api";
import {
  concludedLine,
  countLine,
  engineLine,
  firstSentence,
  freshLine,
  labelsIn,
  modeWords,
  shown,
  shownCount,
  statusWord,
  subjectLine,
  unreadLine,
  weeksOf,
} from "./application";

/**
 * What the page says about a payload, as sentences rather than as a shape.
 *
 * Every date below is written without a zone, so it is read in the machine's
 * own time the way a browser reads one, and the weeks come out the way a human
 * in front of the page would count them.
 */

function landed(over: Partial<Landed> = {}): Landed {
  return {
    id: "g1-s48",
    intent: "The Decisions inbox is done",
    concluded: "landed in 9017baa. The inbox groups collapse and one opens.",
    doneAt: "2026-09-24T09:00:00",
    labels: [],
    arc: "",
    new: false,
    ...over,
  };
}

function page(over: Partial<Page> = {}): Page {
  return {
    schemaVersion: 1,
    readAt: "2026-09-25T11:00:00",
    subject: "MetaSystem",
    mode: "self-hosted",
    engine: { build: "5b9d958", generation: 4, publishedAt: "2026-09-25T10:45:00" },
    counts: { landed: 429, thisMonth: 320, new: 7 },
    landed: [landed()],
    problems: { columns: [], open: [], concluded: [], unread: 0, defects: [], register: "memory/known-issues.md" },
    docs: [],
    visit: { since: "2026-09-24T11:00:00", first: false },
    ...over,
  };
}

describe("the header", () => {
  it("names the subject and the mode, and says the self-hosted case in words", () => {
    expect(subjectLine(page())).toBe("MetaSystem · self-hosted");
    expect(modeWords(page())).toContain("the instance doing the work is on Fleet");
  });

  it("names an adopted application as one built with MetaSystem, and says nothing more", () => {
    const adopted = page({ subject: "ledgerly", mode: "adopted" });
    expect(subjectLine(adopted)).toBe("ledgerly · built with MetaSystem");
    expect(modeWords(adopted)).toBe("");
  });

  // "Last published" is the whole of the honesty: presence is a record at one
  // tick, and this page never claims it is what is running.
  it("says the engine build is the last one published, with its tick", () => {
    expect(engineLine(page())).toBe(
      "last published MetaSystem engine 5b9d958 · generation 4 · published 2026-09-25 10:45",
    );
  });

  it("says so in words where this seat has published none", () => {
    expect(engineLine(page({ engine: null }))).toBe("no engine build available");
    expect(engineLine(page({ engine: { build: "", generation: 0, publishedAt: "" } }))).toBe(
      "no engine build available",
    );
  });

  // "Goals concluded", not "landed": a conclusion records an end, and some of
  // those ends are administrative.
  it("counts the history in the words a conclusion earns", () => {
    expect(countLine(page())).toBe("429 goals concluded · 320 this month · 7 since your last visit");
  });

  it("leaves out a count that has nothing to say", () => {
    expect(countLine(page({ counts: { landed: 12, thisMonth: 0, new: 0 } }))).toBe("12 goals concluded");
  });
});

describe("the weeks", () => {
  const rows = [
    landed({ id: "a", doneAt: "2026-09-25T09:00:00", new: true }),
    landed({ id: "b", doneAt: "2026-09-21T09:00:00" }),
    landed({ id: "c", doneAt: "2026-09-20T09:00:00" }),
    landed({ id: "d", doneAt: "2026-09-14T09:00:00" }),
    landed({ id: "e", doneAt: "" }),
  ];

  // A week is Monday to Sunday: the 20th is a Sunday and belongs to the week
  // that began on the 14th, not to the one starting the next morning.
  it("groups Monday to Sunday, in the order the server sent the rows", () => {
    const weeks = weeksOf(rows);
    expect(weeks.map((week) => week.title)).toEqual([
      "Week of 21 September 2026",
      "Week of 14 September 2026",
      "Nothing dated these",
    ]);
    expect(weeks.map((week) => week.count)).toEqual([2, 2, 1]);
    expect(weeks[0].rows.map((row) => row.id)).toEqual(["a", "b"]);
    expect(weeks[1].rows.map((row) => row.id)).toEqual(["c", "d"]);
  });

  it("counts how many of a week concluded since the last visit, and says nothing where none did", () => {
    const weeks = weeksOf(rows);
    expect(freshLine(weeks[0])).toBe("1 new");
    expect(freshLine(weeks[1])).toBe("");
  });

  it("has no weeks at all where nothing concluded", () => {
    expect(weeksOf([])).toEqual([]);
  });
});

describe("narrowing", () => {
  const rows = [
    landed({ id: "g1-s48", intent: "The Decisions inbox is done", concluded: "landed in 9017baa", labels: ["browser-interface"] }),
    landed({ id: "g1-s42", intent: "The fleet page reads a chain", concluded: "landed in 44de0a1", labels: ["headless-fleet", "robustness"] }),
    landed({ id: "g1-s41", intent: "The census answers the page", concluded: "Obsolete: a duplicate holding no work", labels: ["headless-fleet"] }),
  ];

  it("finds over the id, the intent and the conclusion", () => {
    expect(shown(rows, { find: "g1-s42", label: "" }).map((row) => row.id)).toEqual(["g1-s42"]);
    expect(shown(rows, { find: "inbox", label: "" }).map((row) => row.id)).toEqual(["g1-s48"]);
    expect(shown(rows, { find: "44de0a1", label: "" }).map((row) => row.id)).toEqual(["g1-s42"]);
    expect(shown(rows, { find: "nothing here", label: "" })).toEqual([]);
  });

  it("narrows to one label", () => {
    expect(shown(rows, { find: "", label: "headless-fleet" }).map((row) => row.id)).toEqual(["g1-s42", "g1-s41"]);
  });

  it("draws the chips from the rows shown, commonest first", () => {
    expect(labelsIn(rows)).toEqual([
      { label: "headless-fleet", count: 2 },
      { label: "browser-interface", count: 1 },
      { label: "robustness", count: 1 },
    ]);
  });

  it("says how many are shown of how many there are, and only where they differ", () => {
    expect(shownCount(30, 30)).toBe("30 concluded");
    expect(shownCount(30, 4)).toBe("30 concluded · 4 shown");
  });
});

describe("a row's one line", () => {
  it("is the conclusion's first statement, without the rest of it", () => {
    expect(firstSentence("landed in 9017baa. The inbox groups collapse.")).toBe("landed in 9017baa");
    expect(firstSentence("Obsolete: self-declared duplicate holding no work")).toBe(
      "Obsolete: self-declared duplicate holding no work",
    );
    expect(firstSentence("")).toBe("");
  });
});

describe("known problems", () => {
  function problem(over: Partial<Problem> = {}): Problem {
    return {
      id: "KI-2", date: "2026-08-04", what: "The suite's wall time grew", consequence: "CI cost",
      lever: "Split fast from slow", status: "ACCEPTED 2026-08-06 with a measured trigger", open: false,
      ...over,
    };
  }

  // An accepted limitation still exists, so the word travels onto the line:
  // "ACCEPTED" and "FIXED" are two different things to know about a problem.
  it("keeps the status word visible on a concluded row", () => {
    expect(statusWord(problem())).toBe("ACCEPTED");
    expect(statusWord(problem({ status: "FIXED 2026-08-06: the reader walks the tree" }))).toBe("FIXED");
    expect(statusWord(problem({ status: "RETIRED, superseded by KI-40" }))).toBe("RETIRED");
  });

  it("says how many are concluded on the disclosure that opens them", () => {
    expect(concludedLine([problem(), problem({ id: "KI-3" })])).toBe("Show concluded (2)");
  });

  // Five of this register's own open rows carry three or four cells today, and
  // a block that dropped them would claim the project knows about fewer
  // problems than it does.
  it("counts the rows the reader could not read, in its own line", () => {
    expect(unreadLine(5)).toBe("5 issues could not be read as rows");
    expect(unreadLine(1)).toBe("1 issue could not be read as a row");
    expect(unreadLine(0)).toBe("");
  });
});

File: metasystem/internal/ui/application/application_test.go
package application

import (
	"testing"
	"time"

	"github.com/widoriezebos/agentic-tools/metasystem/internal/backlog"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/knownissues"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/testutil"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/ui/project"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/ui/workspace"
)

// The clock every test below composes against, and the window "new" is decided
// against. Both are injected: nothing here reads a wall clock.
var (
	readAt = time.Date(2026, 9, 25, 11, 0, 0, 0, time.UTC)
	window = time.Date(2026, 9, 24, 11, 0, 0, 0, time.UTC)
)

func done(id, intent, concluded, doneAt string) backlog.Row {
	row := backlog.Row{
		Where: "archived", Lane: backlog.LaneDone, State: "done",
		Intent: intent, Concluded: concluded, DoneAt: doneAt,
	}
	row.ID = id
	return row
}

// The closed rows a board projection hands over: six concluded goals at
// different ages, one of them undated, and one abandoned goal, which is not
// something that concluded.
func closedRows() []backlog.Row {
	rows := []backlog.Row{
		done("g1-s40", "The queue row opens in place", "landed in 7b1c9de; the row opens on the board", "2026-09-25T09:00:00Z"),
		done("g1-s39", "The header counts what is asked", "landed in 2ad91f0", "2026-09-24T09:00:00Z"),
		done("g1-s38", "The label chips are drawn from the rows", "landed in 9cc2e31", "2026-09-18T09:00:00Z"),
		done("g1-s30", "The fleet page reads a seat's chain", "landed in 44de0a1", "2026-08-30T09:00:00Z"),
		done("g1-s29", "A second bundler beside the first", "Obsolete: self-declared duplicate holding no work", ""),
		done("g1-s31", "The stream reconnects by itself", "landed in 1f0aa54", "2026-09-25T09:00:00Z"),
	}
	dropped := backlog.Row{Where: "archived", Lane: backlog.LaneAbandoned, State: "abandoned",
		Intent: "A second bundler", Concluded: "overtaken by g1-s8", DoneAt: "2026-09-25T10:00:00Z"}
	dropped.ID = "g1-s7"
	rows = append(rows, dropped)
	// Labels and an arc, so the row carries what an open row shows.
	rows[0].Labels = []string{"browser-interface", "robustness"}
	rows[0].Arc = "covenant-harvest"
	return rows
}

func composed(over func(*Inputs)) Page {
	in := Inputs{
		Subject: "MetaSystem", Mode: workspace.ModeSelfHosted,
		Engine: &Engine{Build: "5b9d958", Generation: 4, PublishedAt: "2026-09-25T10:45:00Z"},
		Closed: closedRows(),
		// The self-hosted layout: the repository's own README at the checkout
		// root, and the MetaSystem's documents under the installation. Neither
		// root carries a concepts document, so the rule that only a listed
		// path is linked still bites.
		Documents: []project.File{
			{Path: "README.md", Title: "Agentic tools"},
			{Path: "metasystem/README.md", Title: "MetaSystem"},
			{Path: "metasystem/docs/glossary.md", Title: "Glossary"},
			{Path: "metasystem/docs/journey.md", Title: "The journey"},
		},
		Installation: "metasystem",
		RegisterPath: "metasystem/memory/known-issues.md",
		Since:        window,
	}
	if over != nil {
		over(&in)
	}
	return Compose(in, readAt)
}

func landedIDs(page Page) []string {
	ids := []string{}
	for _, row := range page.Landed {
		ids = append(ids, row.ID)
	}
	return ids
}

// Newest first, the undated last, and the abandoned goal is not here at all: a
// goal this project decided not to do is not something the application has.
func TestComposeOrdersTheConcludedGoalsNewestFirstWithTheUndatedLast(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	testutil.Expect(t, "the order", landedIDs(page),
		[]string{"g1-s31", "g1-s40", "g1-s39", "g1-s38", "g1-s30", "g1-s29"})
}

// Two conclusions written in the same second keep one order between reads, so
// a page that is read twice does not shuffle its own rows.
func TestComposeBreaksATieByIDSoTheOrderIsStable(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	testutil.Expect(t, "the two written at 09:00 on the 25th",
		[]string{page.Landed[0].ID, page.Landed[1].ID}, []string{"g1-s31", "g1-s40"})
}

func TestComposeCarriesTheLedgersOwnWordsOnARow(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	testutil.Expect(t, "the newest but one row", page.Landed[1], Landed{
		ID: "g1-s40", Intent: "The queue row opens in place",
		Concluded: "landed in 7b1c9de; the row opens on the board",
		DoneAt:    "2026-09-25T09:00:00Z",
		Labels:    []string{"browser-interface", "robustness"},
		Arc:       "covenant-harvest", New: true,
	})
}

// The header's line: the whole history, this month of it, and what landed
// since this human last read THIS page.
func TestComposeCountsTheHistoryTheMonthAndTheWindow(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	// Two are new: the window opened at 11:00 on the 24th, and the row
	// concluded at 09:00 that same day is on the window's own day but before
	// it, which is an instant compared as an instant rather than as a day.
	testutil.Expect(t, "the counts", page.Counts, Counts{Landed: 6, ThisMonth: 4, New: 2})
}

// A conclusion nothing dated is never new: the record does not say when it was
// written, and a page that guessed would be marking rows new on no evidence.
func TestComposeNeverMarksAnUndatedConclusionNew(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	testutil.Expect(t, "the undated row", page.Landed[5].New, false)
}

// And a page composed over no window at all has nothing new on it, rather than
// every row the ledger ever recorded.
func TestComposeMarksNothingNewWithoutAWindow(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) { in.Since = time.Time{} })
	testutil.Expect(t, "how many are new", page.Counts.New, 0)
	testutil.Expect(t, "the window it says it used", page.Visit, Visit{Since: "", First: false})
}

func TestComposeSaysWhichWindowItDecidedNewAgainst(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) { in.First = true })
	testutil.Expect(t, "the window", page.Visit, Visit{Since: "2026-09-24T11:00:00Z", First: true})
}

func TestComposeNamesTheSubjectTheModeAndTheEngineItWasGiven(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	testutil.Expect(t, "the subject", page.Subject, "MetaSystem")
	testutil.Expect(t, "the mode", page.Mode, "self-hosted")
	testutil.Expect(t, "the engine", page.Engine,
		&Engine{Build: "5b9d958", Generation: 4, PublishedAt: "2026-09-25T10:45:00Z"})
	testutil.Expect(t, "when it was read", page.ReadAt, "2026-09-25T11:00:00Z")
	testutil.Expect(t, "the schema", page.SchemaVersion, SchemaVersion)
}

// A seat that has published no presence has no build to name, and the page
// carries the absence rather than an empty build a human would read as one.
func TestComposeCarriesNoEngineWhereThisSeatHasPublishedNone(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) { in.Engine = nil })
	if page.Engine != nil {
		t.Fatalf("a seat with no presence record has no engine build: %+v", page.Engine)
	}
}

func TestComposeCarriesTheRegisterAsItWasRead(t *testing.T) {
	t.Parallel()
	open := knownissues.Row{ID: "KI-19", Status: "OPEN", Open: true}
	closed := knownissues.Row{ID: "KI-1", Status: "FIXED 2026-08-06"}
	page := composed(func(in *Inputs) {
		in.Register = knownissues.Register{
			Columns:   []string{"Id", "Date", "Issue", "Consequence", "Reopen when", "Status"},
			Open:      []knownissues.Row{open},
			Concluded: []knownissues.Row{closed},
			Unread:    5,
			Defects:   []string{"row=19: wrong column count: got 3, want 6"},
		}
	})
	testutil.Expect(t, "the problems", page.Problems, Problems{
		Columns:   []string{"Id", "Date", "Issue", "Consequence", "Reopen when", "Status"},
		Open:      []knownissues.Row{open},
		Concluded: []knownissues.Row{closed},
		Unread:    5,
		Defects:   []string{"row=19: wrong column count: got 3, want 6"},
		Register:  "metasystem/memory/known-issues.md",
	})
}

// Every list is written whether the register had anything to say or not, so a
// reader never has to tell an absent list from an empty one.
func TestComposeWritesEveryListOfAnEmptyRegister(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) { in.RegisterPath = "" })
	testutil.Expect(t, "the problems", page.Problems, Problems{
		Columns: []string{}, Open: []knownissues.Row{}, Concluded: []knownissues.Row{},
		Unread: 0, Defects: []string{}, Register: "memory/known-issues.md",
	})
}

// The documents this page links, in its own order, and only the ones the
// Project reader lists: an application with no concepts document gets no link
// to one.
//
// Self-hosted, the subject is the MetaSystem, and its README, concepts and
// glossary are the installation's. The checkout's own root README is a
// document about the repository that carries the installation, and it is not
// linked here.
func TestComposeLinksTheInstallationsDocumentsWhereTheWorkspaceIsSelfHosted(t *testing.T) {
	t.Parallel()
	page := composed(nil)
	testutil.Expect(t, "the links", page.Docs, []Document{
		{Title: "README", Path: "metasystem/README.md"},
		{Title: "Glossary", Path: "metasystem/docs/glossary.md"},
	})
}

// Adopted, the subject is the application the checkout holds, so its documents
// are the checkout root's. The kit's own glossary under the installation is
// the machinery's, and this page is not about the machinery.
func TestComposeLinksTheApplicationRootsDocumentsWhereTheWorkspaceIsAdopted(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) {
		in.Mode = workspace.ModeAdopted
		in.Documents = []project.File{
			{Path: "README.md", Title: "The application"},
			{Path: "docs/concepts.md", Title: "Concepts"},
			{Path: "metasystem/docs/glossary.md", Title: "Glossary"},
		}
	})
	testutil.Expect(t, "the links", page.Docs, []Document{
		{Title: "README", Path: "README.md"},
		{Title: "Concepts", Path: "docs/concepts.md"},
	})
}

// A self-hosted layout whose checkout and installation are one directory — the
// kit's own, and the walkthrough's — links the root paths, because that is
// where the installation is.
func TestComposeLinksTheRootPathsWhereTheInstallationIsTheCheckout(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) {
		in.Installation = "."
		in.Documents = []project.File{{Path: "README.md", Title: "MetaSystem"}}
	})
	testutil.Expect(t, "the links", page.Docs, []Document{{Title: "README", Path: "README.md"}})
}

func TestComposeLinksNothingWhereTheCheckoutHasNoneOfThem(t *testing.T) {
	t.Parallel()
	page := composed(func(in *Inputs) { in.Documents = nil })
	testutil.Expect(t, "the links", page.Docs, []Document{})
}

// The month is the observing clock's own UTC month, because the dates are UTC
// instants: a boundary read in another zone would move rows between two
// answers about the same ledger.
func TestComposeCountsTheMonthAgainstTheObservingClock(t *testing.T) {
	t.Parallel()
	in := Inputs{Closed: closedRows(), Since: window}
	october := Compose(in, time.Date(2026, 10, 1, 0, 30, 0, 0, time.UTC))
	testutil.Expect(t, "how many concluded in October", october.Counts.ThisMonth, 0)
}
